File: pipeline/cube_builders/veil.py
"""Veil-of-darkness cube builder.

Reads ``car_ped_stops_veil`` and emits a descriptive cube plus a ``models``
block holding fitted regression results.

Both ``n_stops`` and ``n_motorists`` are carried because the paper's figures
use different denominators: Tables 1-2 are per stop, Figures 2-3 per
motorist, Figure 4 per occupant. A chart that mixes them up silently
reports the wrong number -- Figure 3's headline is 25.1% per motorist but
14.2% per stop.
"""

import sqlite3

import pandas as pd

from build_districts import build as build_district_demographics

from veil.intraracial import (
    DISTRICTS,
    OUTCOMES,
    TREND_DEFAULT_EXCLUDED_YEARS,
    TREND_OUTCOMES,
    TREND_WINDOW,
    TREND_YEARS,
    WINDOW,
    build_sample,
)
from veil.models import (
    INTRARACIAL_SE_TARGETS,
    INTRARACIAL_TARGETS,
    MODEL_TARGETS,
    confidence_interval,
    fit_intraracial,
    fit_vod,
    predicted_probabilities,
)
from veil.sample import int_code
from veil.sun import load_sun_times

CUBE_VERSION = 1

DIMENSIONS = ["year", "era", "clock_bin", "lighting", "party_race", "group_travel", "district"]
MEASURES = ["n_stops", "n_motorists", "n_frisked", "n_ticketed"]

BIN_MINUTES = 15
BLACK = "Black - Non-Latino"
WHITE = "White - Non-Latino"

COMPARISON_RACES = {"black": BLACK, "white": WHITE}


def _normalize_district(d) -> str:
    """Zero-pad a district code to two digits: 2.0 -> "02", 12.0 -> "12".

    ``districtoccur`` is stored REAL in SQLite -- pandas infers a numeric
    dtype from the CSV and reads the zero-padded text "02" as 2.0 -- so a
    bare ``str()`` yields "2.0", whose ``isdigit()`` is False. An earlier
    version of this helper tested ``isdigit()`` directly on that string and
    was therefore a no-op, emitting "12.0" into the cube's `district`
    dimension where its sibling builders emit "12". Route through
    ``veil.sample.int_code`` so the float representation is stripped first.
    """
    code = int_code(d)
    return code.zfill(2) if code else ""


def _descriptive_rows(df: pd.DataFrame) -> list[list]:
    df = df.copy()
    df["clock_bin"] = (df.clock_minutes // BIN_MINUTES) * BIN_MINUTES
    df["district"] = df.districtoccur.map(_normalize_district)
    grouped = df.groupby(
        ["year", "era", "clock_bin", "lighting", "party_race", "group_travel", "district"],
        dropna=False,
    ).agg(
        n_stops=("party_size", "size"),
        n_motorists=("party_size", "sum"),
        n_frisked=("n_frisked", "sum"),
        n_ticketed=("n_ticketed", "sum"),
    ).reset_index()

    return [
        [
            int(r.year), str(r.era), int(r.clock_bin), str(r.lighting),
            str(r.party_race), int(r.group_travel), str(r.district),
            int(r.n_stops), int(r.n_motorists), int(r.n_frisked), int(r.n_ticketed),
        ]
        for r in grouped.itertuples()
    ]


def _time_rounding(df: pd.DataFrame) -> dict:
    """How heavily officers round the recorded stop time.

    The page tells readers that logged times are rounded, which is why the
    ~30-minute band between sunset and full dusk is excluded. That claim
    needs a number attached to it, and the number has to come from our own
    data rather than being asserted in the page's voice: `clock_minutes` is
    minute-level, so this cannot be recomputed from the 15-minute
    `clock_bin` dimension and has to be measured here.
    """
    minutes = df.clock_minutes
    return {
        "n": int(len(minutes)),
        "pct_multiple_of_5": float(100 * (minutes % 5 == 0).mean()),
        "pct_multiple_of_15": float(100 * (minutes % 15 == 0).mean()),
    }


def _fit_all(df: pd.DataFrame) -> dict:
    models: dict = {}

    inter = df.copy()
    # `party_is_black`, not `driver_is_black`: no driver is ever identified in
    # this sample (see veil/sample.py). Every occupant of a retained stop is a
    # young man of the same race, so the outcome is a property of the party.
    inter["party_is_black"] = (inter.party_race == BLACK).astype(int)

    black = df[df.party_race == BLACK].copy()
    black["has_black_passenger"] = black.group_travel

    white = df[df.party_race == WHITE].copy()
    white["has_white_passenger"] = white.group_travel

    jobs = [
        ("party_is_black", inter, "party_is_black"),
        ("has_black_passenger", black, "has_black_passenger"),
        ("placebo_white", white, "has_white_passenger"),
    ]
    for name, data, outcome in jobs:
        for full in (False, True):
            result = fit_vod(data, outcome=outcome, full_controls=full)
            key = f"{name}.{result['spec']}"
            result["seasonality_weight"] = False
            result["paper_coef"] = MODEL_TARGETS.get(key)
            models[key] = result
    return models


def _year_estimates(df: pd.DataFrame) -> list[dict]:
    """Fit the four mutually exclusive age/gender outcomes for every year."""
    estimates = []
    for year in TREND_YEARS:
        subset = df[df.year == year]
        for outcome in TREND_OUTCOMES:
            result = fit_intraracial(subset, outcome)
            ci_lo, ci_hi = confidence_interval(result["coef"], result["se"])
            estimates.append({
                "year": int(year),
                "outcome": outcome,
                "coef": result["coef"],
                "se": result["se"],
                "ci_lo": ci_lo,
                "ci_hi": ci_hi,
                "odds_ratio": result["odds_ratio"],
                "p_value": result["p_value"],
                "n": result["n"],
                "converged": result["converged"],
                "marginal_daylight_pct": result["marginal_daylight_pct"],
                "marginal_dark_pct": result["marginal_dark_pct"],
                "marginal_effect_pp": result["marginal_effect_pp"],
                "marginal_effect_se_pp": result["marginal_effect_se_pp"],
                "marginal_ci_lo_pp": result["marginal_ci_lo_pp"],
                "marginal_ci_hi_pp": result["marginal_ci_hi_pp"],
            })
    return estimates


def _comparison_districts() -> dict[str, tuple[str, ...]]:
    """Dashboard-wide >50% White split, derived from the committed Census input."""
    demographics = build_district_demographics()
    majority_white = tuple(sorted(k for k, v in demographics.items() if v["whiteness"] > 50))
    majority_non_white = tuple(sorted(k for k, v in demographics.items() if v["whiteness"] <= 50))
    return {"majority_white": majority_white, "majority_non_white": majority_non_white}


def _by_year(stops: pd.DataFrame, sun: pd.DataFrame) -> dict:
    """Per-calendar-year fits, for the trend chart above the lead section.

    ONE `build_sample` call over the whole TREND_WINDOW, then split by year --
    deliberately not one call per year. `build_sample` derives both the
    inter-twilight window and the seasonality weights from the window it is
    given, so per-year calls would hand each year a slightly different
    time-of-day window, and a year-over-year difference could then reflect
    that moving window rather than a change in who gets stopped.

    Only the four mutually exclusive group outcomes are fitted
    (TREND_OUTCOMES), separately for Black and White motorists in each of
    the dashboard's two district contexts: 12 years x 4 outcomes x 4 strata.

    These samples are NOT the reproduction's sample and their 2025 is a full
    calendar year where the paper's is January-August. The two must never be
    presented as the same quantity -- see the caveats on the page.
    """
    strata = []
    for district_context, districts in _comparison_districts().items():
        for race, race_value in COMPARISON_RACES.items():
            sample = build_sample(
                stops, sun, window=TREND_WINDOW,
                races=(race_value,), districts=districts,
            )
            strata.append({
                "race": race,
                "district_context": district_context,
                "districts": list(districts),
                "estimates": _year_estimates(sample),
            })

    return {
        "window": {"start": TREND_WINDOW[0], "end": TREND_WINDOW[1]},
        "years": [int(y) for y in TREND_YEARS],
        # Which years the chart shows before the reader touches anything.
        # Emitted rather than hardcoded in the page so the exclusion and its
        # reason live in one place, next to the sample that produced it.
        "default_years": [
            int(y) for y in TREND_YEARS if y not in TREND_DEFAULT_EXCLUDED_YEARS
        ],
        "strata": strata,
    }


def _intraracial(conn: sqlite3.Connection) -> dict:
    """Hannon & Biddle (2025) intraracial age/gender models and probabilities.

    Reads the dedicated `car_ped_stops_veil_intraracial` table -- a
    differently-filtered, differently-sampled table than `car_ped_stops_veil`
    above -- and must never have its rows mixed into the 2026 group-travel
    cube's `rows`/`models`/`time_rounding` keys.
    """
    stops = pd.read_sql("SELECT * FROM car_ped_stops_veil_intraracial", conn)
    sun = load_sun_times()
    df = build_sample(stops, sun)

    models = {}
    for outcome in OUTCOMES:
        result = fit_intraracial(df, outcome)
        models[outcome] = {
            "coef": result["coef"],
            "se": result["se"],
            "odds_ratio": result["odds_ratio"],
            "p_value": result["p_value"],
            "n": result["n"],
            "converged": result["converged"],
            "paper_coef": INTRARACIAL_TARGETS.get(outcome),
            "paper_se": INTRARACIAL_SE_TARGETS.get(outcome),
        }

    probabilities = []
    for group in ("young_male", "young_female", "older_male", "older_female"):
        probs = predicted_probabilities(df, group)
        probabilities.append({"group": group, "lighting": "daylight", "pct": probs["daylight"]})
        probabilities.append({"group": group, "lighting": "dark", "pct": probs["dark"]})

    return {
        "sample": {
            "n": int(len(df)),
            "window_start": WINDOW[0],
            "window_end": WINDOW[1],
            "districts": list(DISTRICTS),
        },
        "models": models,
        "probabilities": probabilities,
        # A second, wider sample over the same table: full calendar years
        # 2021-2025, fitted one year at a time. Sibling to the three keys
        # above, never merged into them.
        "by_year": _by_year(stops, sun),
    }


def build(conn: sqlite3.Connection) -> tuple[dict, dict]:
    df = pd.read_sql("SELECT * FROM car_ped_stops_veil", conn)
    if df.empty:
        raise SystemExit(
            "car_ped_stops_veil is empty or missing. Run: python -m veil.build "
            "--zip data/<backup>.zip --db data/<db>.db"
        )

    replication = df[df.year.between(2021, 2024)]
    cube = {
        "version": CUBE_VERSION,
        "dimensions": DIMENSIONS,
        "measures": MEASURES,
        "rows": _descriptive_rows(df),
        "models": _fit_all(replication),
        "time_rounding": _time_rounding(replication),
        "intraracial": _intraracial(conn),
    }
    return cube, {}

File: utils/veil.ts
/**
 * Helpers for the veil-of-darkness cube.
 *
 * Denominators differ per figure and mixing them up silently reports the
 * wrong number. The paper's headline "five times more likely" (25.1% vs
 * 4.6%) is per motorist; the same quantity per stop is 14.2% vs 2.6%.
 * Function names here say which denominator they use.
 *
 * None of the selectors below take a year/date range. They aggregate
 * unconditionally over every row in the `VeilCube` they're given. The
 * shipped cube spans 2014-2026, not just the paper's 2021-2024 study
 * window, so callers must call `restrictToYears(cube, REPLICATION_YEARS.from,
 * REPLICATION_YEARS.to)` once and pass the result to every selector —
 * there is no `opts` parameter on the selectors themselves.
 */
import type { Cube } from '~/utils/cube'

export interface VeilModel {
  coef: number
  se: number
  p: number
  n: number
  odds_ratio: number
  spec: string
  converged: boolean
  seasonality_weight: boolean
  paper_coef: number | null
  /**
   * Share of ROWS whose value for a collapsible fixed effect was folded
   * into the "OTHER" bucket, keyed by column (`police_area`,
   * `assigned_unit`). Only present on model_2 fits. A large share means
   * that control is hollow for that fit and must be disclosed, not shipped
   * silently.
   */
  other_row_share?: Record<string, number>
}

/**
 * How heavily officers round the recorded stop time, measured by the
 * pipeline on the replication slice. `clock_minutes` is minute-level and
 * the cube's `clock_bin` dimension is 15-minute, so these cannot be
 * recomputed from `rows` — they are carried here so the page can state a
 * measured figure instead of an unsourced one.
 */
export interface VeilTimeRounding {
  n: number
  pct_multiple_of_5: number
  pct_multiple_of_15: number
}

/**
 * One outcome's fitted intraracial (Hannon & Biddle 2025) model. Unlike
 * `VeilModel` above, there is no `spec`/`seasonality_weight` split — every
 * intraracial fit applies the full control set and the paper's seasonality
 * weight.
 */
export interface VeilIntraracialModel {
  coef: number
  se: number
  odds_ratio: number
  p_value: number
  n: number
  converged: boolean
  /** Hannon & Biddle (2025) Table 1, "Dark Out" column -- from INTRARACIAL_TARGETS. */
  paper_coef: number | null
  /** Hannon & Biddle (2025) Table 1 standard error -- from INTRARACIAL_SE_TARGETS. */
  paper_se: number | null
}

export type VeilIntraracialGroup = 'young_male' | 'young_female' | 'older_male' | 'older_female'
export type VeilIntraracialLighting = 'daylight' | 'dark'
export type VeilComparisonRace = 'black' | 'white'
export type VeilDistrictContext = 'majority_white' | 'majority_non_white'

export interface VeilIntraracialProbability {
  group: VeilIntraracialGroup
  lighting: VeilIntraracialLighting
  pct: number
}

export interface VeilIntraracialSample {
  n: number
  window_start: string
  window_end: string
  districts: string[]
}

/**
 * One outcome fitted on one calendar year, for the trend chart.
 *
 * `ci_lo`/`ci_hi` are the 95% Wald bounds the pipeline computes; the chart
 * draws them rather than recomputing from `se`, so the whiskers and any
 * table of these numbers can never disagree.
 */
export interface VeilIntraracialYearEstimate {
  year: number
  outcome: VeilIntraracialGroup
  coef: number
  se: number
  ci_lo: number
  ci_hi: number
  odds_ratio: number
  p_value: number
  n: number
  converged: boolean
  /** Sample-average predicted probabilities under the two counterfactuals. */
  marginal_daylight_pct: number
  marginal_dark_pct: number
  /** Dark minus daylight, in percentage points, with delta-method bounds. */
  marginal_effect_pp: number
  marginal_effect_se_pp: number
  marginal_ci_lo_pp: number
  marginal_ci_hi_pp: number
}

/**
 * Year-by-year fits, a SEPARATE sample from the reproduction above.
 *
 * These are full calendar years, where `VeilIntraracialSample` covers
 * January 2022 - August 2025. 2025 therefore appears in both and means
 * something different in each; anything rendering both must say so.
 */
export interface VeilIntraracialByYear {
  window: { start: string; end: string }
  years: number[]
  /**
   * The years shown before the reader changes anything. A subset of `years`:
   * 2020 is fitted and selectable but omitted here, because lockdowns moved
   * evening travel and the veil-of-darkness design assumes travel patterns
   * do not shift with the light.
   */
  default_years: number[]
  strata: Array<{
    race: VeilComparisonRace
    district_context: VeilDistrictContext
    districts: string[]
    estimates: VeilIntraracialYearEstimate[]
  }>
  pooled_race_interaction: VeilPooledRaceInteraction[]
}

export interface VeilPooledRaceEffect {
  daylight_pct: number
  dark_pct: number
  effect_pp: number
  effect_se_pp: number
  effect_ci_lo_pp: number
  effect_ci_hi_pp: number
}

export interface VeilPooledRaceInteraction {
  district_context: VeilDistrictContext
  districts: string[]
  outcome: 'young_male'
  n: number
  converged: boolean
  interaction_coef: number
  interaction_se: number
  interaction_p_value: number
  interaction_odds_ratio: number
  effects: Record<VeilComparisonRace, VeilPooledRaceEffect>
  /** Black after-dark change minus white after-dark change. */
  difference_pp: number
  difference_se_pp: number
  difference_ci_lo_pp: number
  difference_ci_hi_pp: number
  difference_p_value: number
}

export interface VeilMarginalProbabilityAggregate {
  outcome: VeilIntraracialGroup
  n: number
  daylightPct: number
  darkPct: number
  effectPp: number
  effectCiLoPp: number
  effectCiHiPp: number
}

/**
 * Combine annual standardized predictions for a display-only pooled view.
 * Each annual probability is weighted by that fit's stop count. This is not
 * a pooled regression: callers must describe it as an aggregation of annual
 * estimates so it cannot be mistaken for the paper-window model.
 */
export function aggregateMarginalProbabilities(
  estimates: VeilIntraracialYearEstimate[],
  outcomes: VeilIntraracialGroup[],
  years: number[],
): VeilMarginalProbabilityAggregate[] {
  const chosenOutcomes = new Set(outcomes)
  const chosenYears = new Set(years)
  const totals = new Map<VeilIntraracialGroup, {
    n: number; daylight: number; dark: number; effect: number; effectVariance: number
  }>()

  for (const estimate of estimates) {
    if (!chosenOutcomes.has(estimate.outcome) || !chosenYears.has(estimate.year)) continue
    const total = totals.get(estimate.outcome) ?? {
      n: 0, daylight: 0, dark: 0, effect: 0, effectVariance: 0,
    }
    total.n += estimate.n
    total.daylight += estimate.marginal_daylight_pct * estimate.n
    total.dark += estimate.marginal_dark_pct * estimate.n
    total.effect += estimate.marginal_effect_pp * estimate.n
    total.effectVariance += (estimate.n * estimate.marginal_effect_se_pp) ** 2
    totals.set(estimate.outcome, total)
  }

  return outcomes.flatMap((outcome) => {
    const total = totals.get(outcome)
    if (!total?.n) return []
    const daylightPct = total.daylight / total.n
    const darkPct = total.dark / total.n
    const effectPp = total.effect / total.n
    const effectSe = Math.sqrt(total.effectVariance) / total.n
    return [{
      outcome,
      n: total.n,
      daylightPct,
      darkPct,
      effectPp,
      effectCiLoPp: effectPp - 1.96 * effectSe,
      effectCiHiPp: effectPp + 1.96 * effectSe,
    }]
  })
}

export interface VeilIntraracial {
  sample: VeilIntraracialSample
  models: Record<string, VeilIntraracialModel>
  probabilities: VeilIntraracialProbability[]
  by_year: VeilIntraracialByYear
}

export interface VeilCube extends Cube {
  models: Record<string, VeilModel>
  time_rounding?: VeilTimeRounding
  intraracial?: VeilIntraracial
}

type Row = Array<string | number | null>

/**
 * The paper's study window. The shipped cube spans 2014-2026 (it is
 * rebuilt from the full stops history), so every selector in this file
 * aggregates over whatever rows are in the cube it's given — callers
 * MUST restrict to this range themselves (see `restrictToYears`) before
 * calling a selector, or the numbers will not match the paper's figures.
 */
export const REPLICATION_YEARS = { from: 2021, to: 2024 } as const

/**
 * Return a new `VeilCube` whose `rows` are limited to `year` in
 * `[fromYear, toYear]` (inclusive). Does not mutate `cube`.
 * `dimensions`, `measures`, `version`, and `models` pass through
 * unchanged (`models` describes fitted regressions and isn't
 * re-derived here). Compose this once at the call site and pass the
 * result to every selector, rather than threading a year range through
 * each one individually.
 */
export function restrictToYears(cube: VeilCube, fromYear: number, toYear: number): VeilCube {
  const yearIdx = column(cube, 'year')
  return {
    ...cube,
    rows: cube.rows.filter((row) => {
      const y = Number((row as Row)[yearIdx])
      return y >= fromYear && y <= toYear
    }),
  }
}

function column(cube: VeilCube, name: string): number {
  const dim = cube.dimensions.indexOf(name)
  if (dim >= 0) return dim
  const measure = cube.measures.indexOf(name)
  if (measure < 0) throw new Error(`unknown cube column: ${name}`)
  return cube.dimensions.length + measure
}

function sumBy(
  cube: VeilCube,
  measure: string,
  predicate: (get: (name: string) => string | number | null) => boolean,
): number {
  const idx = column(cube, measure)
  let total = 0
  for (const row of cube.rows as Row[]) {
    const get = (name: string) => row[column(cube, name)]
    if (predicate(get)) total += Number(row[idx] ?? 0)
  }
  return total
}

const RACES = ['Black - Non-Latino', 'White - Non-Latino'] as const

/** Total young male motorists stopped, by race. Paper figure 2. */
export function motoristsByRace(cube: VeilCube): Record<string, number> {
  const out: Record<string, number> = {}
  for (const race of RACES) {
    out[race] = sumBy(cube, 'n_motorists', (get) => get('party_race') === race)
  }
  return out
}

/** Share of motorists who were in a multi-occupant car. Paper figure 3. */
export function pctMotoristsInGroups(cube: VeilCube): Record<string, number> {
  const out: Record<string, number> = {}
  for (const race of RACES) {
    const all = sumBy(cube, 'n_motorists', (get) => get('party_race') === race)
    const grouped = sumBy(
      cube,
      'n_motorists',
      (get) => get('party_race') === race && Number(get('group_travel')) === 1,
    )
    out[race] = all === 0 ? 0 : (grouped / all) * 100
  }
  return out
}

/** Per-occupant frisk and ticket rates, solo vs group. Paper figure 4. */
export function frisksAndTickets(cube: VeilCube, race: string) {
  const bucket = (grouped: 0 | 1) => {
    const match = (get: (n: string) => string | number | null) =>
      get('party_race') === race && Number(get('group_travel')) === grouped
    const motorists = sumBy(cube, 'n_motorists', match)
    return {
      motorists,
      friskRate: motorists === 0 ? 0 : (sumBy(cube, 'n_frisked', match) / motorists) * 100,
      ticketRate: motorists === 0 ? 0 : (sumBy(cube, 'n_ticketed', match) / motorists) * 100,
    }
  }
  return { solo: bucket(0), group: bucket(1) }
}

/**
 * The dashboard-wide definition of a majority-white police district:
 * resident whiteness above 50% in `public/cubes/districts.json`. Same rule
 * the neighborhoods page and the disparity sentence use — see
 * `pages/data.vue`'s "How does this dashboard calculate majority-white
 * districts?" entry.
 */
export const MAJORITY_WHITE_THRESHOLD = 50

/**
 * Share of stops of one race that happened in a majority-white police
 * district. Per stop.
 *
 * Depends on the cube's `district` dimension carrying zero-padded two-digit
 * codes ('02', '12') so it can be keyed against `districts.json`. Districts
 * missing from the demographics table (e.g. '77', an airport district with
 * no residents) are excluded from BOTH numerator and denominator, so the
 * result is "of stops in districts we can classify".
 */
export function pctStopsInMajorityWhiteDistricts(
  cube: VeilCube,
  race: string,
  demographics: Record<string, { whiteness: number }>,
): { pct: number; stops: number; classifiedStops: number } {
  const districtIdx = column(cube, 'district')
  const raceIdx = column(cube, 'party_race')
  const stopsIdx = column(cube, 'n_stops')

  let stops = 0
  let classified = 0
  let inMajorityWhite = 0
  for (const row of cube.rows as Row[]) {
    if (row[raceIdx] !== race) continue
    const n = Number(row[stopsIdx] ?? 0)
    stops += n
    const entry = demographics[String(row[districtIdx])]
    if (!entry) continue
    classified += n
    if (entry.whiteness > MAJORITY_WHITE_THRESHOLD) inMajorityWhite += n
  }
  return {
    pct: classified === 0 ? 0 : (inMajorityWhite / classified) * 100,
    stops,
    classifiedStops: classified,
  }
}

/** Display order for lighting states: daylight side of the veil first. */
export const LIGHT_ORDER = ['daylight', 'dark'] as const

export interface ClockBinPoint {
  clockBin: number
  lighting: string
  pctGroupStops: number
  /**
   * Stops behind `pctGroupStops` — the denominator, not a rate. The window
   * is truncated at both ends (17:08-20:35), so the first dark bin and the
   * last daylight bin rest on very few stops and their percentages swing
   * wildly. Callers plotting these points must use this to suppress thin
   * bins rather than letting one 40-stop bin set a chart's y-axis.
   */
  stops: number
}

/**
 * Share of stops that are group-travel, per clock-time bin and lighting
 * state. Per stop, not per motorist: this mirrors the regression outcome.
 * The gap between the two lines at the same clock time is the veil-of-
 * darkness contrast made visible.
 */
export function groupTravelByClockBin(cube: VeilCube, race: string): ClockBinPoint[] {
  const points = new Map<string, { total: number; grouped: number; bin: number; light: string }>()
  const binIdx = column(cube, 'clock_bin')
  const lightIdx = column(cube, 'lighting')
  const raceIdx = column(cube, 'party_race')
  const groupIdx = column(cube, 'group_travel')
  const stopsIdx = column(cube, 'n_stops')

  for (const row of cube.rows as Row[]) {
    if (row[raceIdx] !== race) continue
    const bin = Number(row[binIdx])
    const light = String(row[lightIdx])
    const key = `${bin}|${light}`
    const entry = points.get(key) ?? { total: 0, grouped: 0, bin, light }
    const n = Number(row[stopsIdx] ?? 0)
    entry.total += n
    if (Number(row[groupIdx]) === 1) entry.grouped += n
    points.set(key, entry)
  }

  return [...points.values()]
    .sort((a, b) => a.bin - b.bin || LIGHT_ORDER.indexOf(a.light) - LIGHT_ORDER.indexOf(b.light))
    .map((e) => ({
      clockBin: e.bin,
      lighting: e.light,
      pctGroupStops: e.total === 0 ? 0 : (e.grouped / e.total) * 100,
      stops: e.total,
    }))
}

/** Why a clock bin was dropped from chart 4. Prose is the caller's business. */
export type SuppressedBin =
  | { clockBin: number; reason: 'unpaired'; present: string[] }
  | { clockBin: number; reason: 'thin'; thin: Array<{ lighting: string; stops: number }> }

export interface ClockBinSelection {
  /** Kept bins, ascending by clock time, each holding both lighting states. */
  kept: ClockBinPoint[][]
  /** Dropped bins, ascending by clock time, one entry per bin. */
  suppressed: SuppressedBin[]
  /** Kept bins where the daylight share exceeds the dark one. */
  daylightHigherBins: number
}

/**
 * Choose which clock bins chart 4 may draw.
 *
 * A bin is a veil-of-darkness comparison only if it is observed in BOTH
 * lighting states and both rest on enough stops. The window is truncated at
 * both ends (17:08-20:35), so the first dark bin and the last daylight bin sit
 * on very few stops and their percentages swing wildly; one 40-stop bin would
 * otherwise set the chart's y-axis.
 *
 * Bins are kept or dropped as a pair. Half a pair reads as a missing
 * comparison rather than a thin one, and it corrupts the counts: a lone bar
 * would be counted in `kept` yet, having no counterpart to beat, never counted
 * in `daylightHigherBins`, quietly deflating the "12 of the 13 bins" sentence
 * on the page. Testing thickness alone does not catch that — `some`/`every`
 * over a one-element array is vacuous — which is why the pairing check is
 * separate and comes first. No such bin exists in the current cube; this closes
 * the case rather than relying on that staying true.
 *
 * `minStops` is inclusive: a side resting on exactly the minimum is thick
 * enough.
 */
export function selectClockBinPairs(
  points: ClockBinPoint[],
  minStops: number,
): ClockBinSelection {
  const bins = new Map<number, ClockBinPoint[]>()
  for (const p of points) {
    const existing = bins.get(p.clockBin)
    if (existing) existing.push(p)
    else bins.set(p.clockBin, [p])
  }

  const kept: ClockBinPoint[][] = []
  const suppressed: SuppressedBin[] = []

  for (const [clockBin, binPoints] of [...bins.entries()].sort((a, b) => a[0] - b[0])) {
    const missing = LIGHT_ORDER.filter((state) => !binPoints.some((p) => p.lighting === state))
    if (missing.length > 0) {
      // Unpaired is reported ahead of thinness: a lone bar has no comparison to
      // be thin about, and naming both reasons would overstate the case.
      suppressed.push({
        clockBin,
        reason: 'unpaired',
        present: LIGHT_ORDER.filter((state) => binPoints.some((p) => p.lighting === state)),
      })
      continue
    }
    const thin = binPoints.filter((p) => p.stops < minStops)
    if (thin.length > 0) {
      suppressed.push({
        clockBin,
        reason: 'thin',
        thin: thin.map((p) => ({ lighting: p.lighting, stops: p.stops })),
      })
      continue
    }
    kept.push(binPoints)
  }

  const daylightHigherBins = kept.filter((binPoints) => {
    const day = binPoints.find((p) => p.lighting === 'daylight')
    const dark = binPoints.find((p) => p.lighting === 'dark')
    return day !== undefined && dark !== undefined && day.pctGroupStops > dark.pctGroupStops
  }).length

  return { kept, suppressed, daylightHigherBins }
}
